server/src/main/java/io/crate/execution/dml/Indexer.java

```diff
@@ -175,7 +175,7 @@ public CollectExpression<IndexItem, Object> getImplementation(Reference ref) {
                 int pIndex = table.partitionedByColumns().indexOf(ref);
                 if (pIndex > -1) {
                     String val = partitionName.values().get(pIndex);
-                    return NestableCollectExpression.constant(val);
+                    return NestableCollectExpression.constant(ref.valueType().implicitCast(val));
                 } else {
```

Comment:
Otherwise scalar function representing CHECK fails with `"left and right must have the same type for comparison"`